src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -119,120 +106,285 @@ EmbeddedAssemblies::install_preload_hooks ()
 	mono_install_assembly_refonly_preload_hook (open_from_bundles_refonly, nullptr);
 }
 
-int
-EmbeddedAssemblies::TypeMappingInfo_compare_key (const void *a, const void *b)
+template<typename Key, typename Entry, int (*compare)(const Key*, const Entry*), bool use_extra_size>
+const Entry*
+EmbeddedAssemblies::binary_search (const Key *key, const Entry *base, size_t nmemb, [[maybe_unused]] size_t extra_size)
 {
-	return strcmp (reinterpret_cast <const char*> (a), reinterpret_cast <const char*> (b));
-}
+	static_assert (compare != nullptr, "compare is a required template parameter");
 
-inline const char*
-EmbeddedAssemblies::find_entry_in_type_map (const char *name, uint8_t map[], TypeMapHeader& header)
-{
-	const char *e = reinterpret_cast<const char*> (bsearch (name, map, header.entry_count, header.entry_length, TypeMappingInfo_compare_key ));
-	if (e == nullptr)
+	// This comes from the app, so let's be civil
```

Comment:
if it comes from the app, shouldn't we just SIGSEGV?

...and why am I having a strong sense of deja vu writing that comment?

Reply:
...perhaps the issue is I'm misinterpreting "the app", and this could be "not us user code"?

Reply:
The "user code" ("something outside XA runtime") may, somehow, pass a null string to e.g. `typemap_java_to_managed` and I see no reason to crash in this case. If "they" want to crash, that's OK, but since it's an outside factor from our POV, we should recover gracefully.